modules/artifact-activation/src/main/java/org/mule/runtime/module/artifact/activation/internal/descriptor/AbstractArtifactDescriptorFactory.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.artifact.activation.internal.descriptor;
+
+import org.mule.runtime.api.deployment.meta.AbstractMuleArtifactModel;
+import org.mule.runtime.api.meta.MuleVersion;
+import org.mule.runtime.module.artifact.api.descriptor.ArtifactDescriptor;
+import org.mule.runtime.module.artifact.api.descriptor.ArtifactDescriptorValidator;
+import org.mule.runtime.module.artifact.api.descriptor.ArtifactDescriptorValidatorBuilder;
+import org.mule.runtime.module.artifact.api.descriptor.BundleDescriptor;
+import org.mule.runtime.module.artifact.api.descriptor.ClassLoaderModel;
+
+import java.io.File;
+
+/**
+ * Base class to create artifact descriptors.
+ *
+ * @param <M> type of the artifact model that owns the descriptor.
+ * @param <T> type of descriptor being created.
+ */
+public abstract class AbstractArtifactDescriptorFactory<M extends AbstractMuleArtifactModel, T extends ArtifactDescriptor> {
+
+  protected final File artifactLocation;
+  protected final M artifactModel;
+  private final ArtifactDescriptorValidator artifactDescriptorValidator;
+
+  public AbstractArtifactDescriptorFactory(File artifactLocation, M artifactModel,
+                                           ArtifactDescriptorValidatorBuilder artifactDescriptorValidatorBuilder) {
+    this.artifactLocation = artifactLocation;
+    this.artifactModel = artifactModel;
+
+    this.artifactDescriptorValidator = artifactDescriptorValidatorBuilder
+        .validateMinMuleVersion()
+        .validateMuleProduct()
+        .validateVersionFormat()
+        .build();
+  }
+
+  public T createArtifactDescriptor() {
+    final T descriptor = doCreateArtifactDescriptor();
+    if (artifactLocation.isDirectory()) {
+      descriptor.setRootFolder(artifactLocation);
+    }
+
+    BundleDescriptor bundleDescriptor = getBundleDescriptor();
+    descriptor.setBundleDescriptor(bundleDescriptor);
+    descriptor.setMinMuleVersion(new MuleVersion(artifactModel.getMinMuleVersion()));
+    descriptor.setRequiredProduct(artifactModel.getRequiredProduct());
+
+    ClassLoaderModel classLoaderModel =
+        getClassLoaderModel();
+    descriptor.setClassLoaderModel(classLoaderModel);
+
+    doDescriptorConfig(descriptor);
+
+    // TODO: uncomment once the model is guaranteed to be completed with all the necessary information
```

Comment:
what is missing to guarantee that?

Reply:
Doing what the packager does with `DefaultValuesMuleArtifactJsonGenerator` class.

Reply:
Created user story [W-11203071](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000xoVUsYAM/view) to tackle it.